app/src/main/kotlin/com/exactpro/th2/lwdataprovider/grpc/GrpcDataProviderBackPressure.kt

```diff
@@ -57,50 +58,90 @@ class GrpcDataProviderBackPressure(
-            while (servCallObs.isReady && inProcess) {
+            while (servCallObs.isReady && inProcess && !isCancelled.get()) {
                 if (servCallObs.isCancelled) {
                     logger.warn { "Request is canceled during processing" }
-                    handler.cancel()
+                    cancel()
                     return@setOnReadyHandler
                 }
-                val event = buffer.take()
-                if (event.close) {
-                    servCallObs.onCompleted()
-                    inProcess = false
-                    onFinished()
-                    onClose(handler)
-                    logger.info { "Executing finished successfully" }
-                } else if (event.error != null) {
-                    servCallObs.onError(event.error)
-                    inProcess = false
-                    onFinished()
-                    handler.complete()
-                    logger.warn(event.error) { "Executing finished with error" }
-                } else {
-                    converter.invoke(event)?.let {  servCallObs.onNext(it) }
+
+                try {
+                    // We need to poll because if we will use take and keepOpen option it is possible that we will have to wait here indefinitely
+                    val event = buffer.poll(EVENT_POLLING_TIMEOUT, TimeUnit.MILLISECONDS) ?: continue
+                    when {
+                        event.close -> {
+                            servCallObs.onCompleted()
+                            inProcess = false
+                            onFinished()
+                            onClose(handler)
+                            logger.info { "Executing finished successfully" }
+                        }
+                        event.error != null -> {
+                            servCallObs.onError(event.error)
+                            inProcess = false
+                            onFinished()
+                            handler.complete()
+                            logger.warn(event.error) { "Executing finished with error" }
+                        }
+                        else -> {
+                            converter.invoke(event)?.let { servCallObs.onNext(it) }
+                        }
+                    }
+                } catch (e: InterruptedException) {
+                    logger.warn(e) { "Processing interrupted" }
+                    cancel()
+                    return@setOnReadyHandler
+                } catch (e: Exception) {
+                    logger.error(e) { "Error processing event" }
+                    servCallObs.onError(Status.INTERNAL
+                        .withDescription("Internal error during processing")
+                        .withCause(e)
+                        .asRuntimeException())
+                    cancel()
+                    return@setOnReadyHandler
                 }
             }
-            if (inProcess) {
+
+            if (inProcess && !isCancelled.get()) {
```

Comment:
question: Same as above about `handler.isAlive` usage. Can we replace `isCancelled`?